### PR TITLE
ARROW-2920: [Python] Fix pytorch segfault

### PR DESCRIPTION
**Pull request by pcmoritz**
This fixes ARROW-2920 (see also https://github.com/ray-project/ray/issues/2447) for me

Unfortunately we might not be able to have regression tests for this right now because we don't have CUDA in our test toolchain.

### CODE REVIEW

**Comment by pcmoritz** [comment]
Before we merge this, let me add the regression test (skipping it) even though we can't run it in the CI.

**Comment by pcmoritz** [comment]
+1 This is ready to merge from my side

File: python/pyarrow/compat.py [hunk 1]
```diff
@@ -205,16 +191,37 @@ def import_tensorflow_extension():
...
+            yield path
 
+def import_tensorflow_extension():
+    """
+    Load the TensorFlow extension if it exists.
+
+    This is used to load the TensorFlow extension before
+    pyarrow.lib. If we don't do this there are symbol clashes
+    between TensorFlow's use of threading and our global
+    thread pool, see also
```
Comment (wesm) [inline python/pyarrow/compat.py]: The issue is a bit more complicated than that. The problem (is my understanding) is that devtoolset-2 statically links various parts of the C and C++ runtime, including some mutable static data members which are modified by pthreads initialization. We should document this issue in some more detail somewhere so we can point packaging folks to it to see what can be done to improve the situation in the future

**Review by wesm — APPROVED** [review]
+1

**Comment by ppwwyyxx** [comment]
Why are we using this kind of hacks again and again?
I've mentioned in https://github.com/apache/arrow/pull/1223#issuecomment-359892433 about the downsides of importing a giant library like pytorch and tensorflow.
Importing them may appear to solve some symbol collision issue, but in the long run it only makes things worse by bringing in more and more symbols from all the libraries.

I recently found that pyarrow appears to also have symbol collision issues with [horovod](https://github.com/uber/horovod/). Do you want to also add `compat.import_horovod_extension()`?
I hope not. 

`readelf -Ws libarrow.so.10 | grep 'WEAK.*DEFAULT' | grep -v 'arrow'` is giving me many symbols that should not appear: flatbuffers, protobuf, rapidjson....  I think that's where you want to look at

**Comment by wesm** [comment]
@ppwwyyxx your comment does not seem constructive to me. Would you like to help remedy the situation?

**Pull request closed without merging**